Add cmake support (#528)
libtom/libtommath: CMakeLists.txt

# integer (MPI) library written entirely in C.
#

cmake_minimum_required(VERSION 3.10)

project(tommath VERSION 1.2.0)

# package release version
# bump if re-releasing the same VERSION + patches
# set to 1 if releasing a new version
set(PACKAGE_RELEASE_VERSION 1)

#-----------------------------------------------------------------------------
# Include cmake modules
#-----------------------------------------------------------------------------
include(GNUInstallDirs)
include(CMakePackageConfigHelpers)
include(sources.cmake)

# The only direct cmake argument for now
option(BUILD_SHARED_LIBS "Build shared library and only the shared library if \"ON\", default is static" OFF)

#-----------------------------------------------------------------------------
# Compose CFLAGS
#-----------------------------------------------------------------------------

# Some information copied from makefile_include.mk

# Basic set
set(LTM_C_FLAGS  -Wall -Wsign-compare -Wextra -Wshadow)
set(LTM_C_FLAGS  ${LTM_C_FLAGS} -Wdeclaration-after-statement -Wbad-function-cast -Wcast-align)
set(LTM_C_FLAGS  ${LTM_C_FLAGS} -Wstrict-prototypes -Wpointer-arith -Wsystem-headers)

> if(COMPILE_DEBUG)

Author: Probably [CMAKE_BUILD_TYPE](https://cmake.org/cmake/help/v3.10/variable/CMAKE_BUILD_TYPE.html#cmake-build-type) should be used for this kind of thing.

CMAKE_C_FLAGS_[DEBUG|RELEASE|RELWITHDEBINFO|MINSIZEREL] variables are used depending on the build type with the following default values

```
CMAKE_C_FLAGS_DEBUG: -g
CMAKE_C_FLAGS_RELEASE: -O3 -DNDEBUG
CMAKE_C_FLAGS_MINSIZEREL: -Os -DNDEBUG
CMAKE_C_FLAGS_RELWITHDEBINFO: -O2 -g -DNDEBUG
```

Default values can be overriden e.g. for -g3 in debug builds `set(CMAKE_C_FLAGS_DEBUG -g3)`

Reviewer: Fine like that?

# SPDX-License-Identifier: Unlicense
#
# LibTomMath, a free open source portable number theoretic multiple-precision
# integer (MPI) library written entirely in C.
#

cmake_minimum_required(VERSION 3.10)

project(tommath
    VERSION 1.2.0
    DESCRIPTION "A free open source portable number theoretic multiple-precision integer (MPI) library written entirely in C."
    HOMEPAGE_URL "https://www.libtom.net/LibTomMath"
    LANGUAGES C)

# package release version
# bump if re-releasing the same VERSION + patches
# set to 1 if releasing a new VERSION
set(PACKAGE_RELEASE_VERSION 1)

#-----------------------------------------------------------------------------
# Include cmake modules
#-----------------------------------------------------------------------------
include(GNUInstallDirs)
include(CheckIPOSupported)
include(CMakePackageConfigHelpers)
# default is "No tests"

> option(BUILD_TESTING "" OFF)
> include(CTest)

Reviewer 2: If I read the [`CTest`](https://cmake.org/cmake/help/latest/module/CTest.html) documentation correctly, the `BUILD_TESTING` option is created by the module.
```suggestion
include(CTest)
```

Reviewer: Yes, and as stated in the comment above that `option()` statement we want it to be turned `Off` by default, while CTest declares it as `On`

Reviewer 2: Oops! Sorry! I didn't read that line.

include(sources.cmake)

# The only direct cmake argument for now
option(BUILD_SHARED_LIBS "Build shared library and only the shared library if \"ON\", default is static" OFF)

#-----------------------------------------------------------------------------
# Compose CFLAGS
#-----------------------------------------------------------------------------

# Some information ported from makefile_include.mk

if(NOT CMAKE_BUILD_TYPE AND NOT CMAKE_CONFIGURATION_TYPES)
  message(STATUS "Setting build type to 'Release' as none was specified.")

>   set(CMAKE_BUILD_TYPE "Release")

Reviewer 2: ```suggestion
  set(CMAKE_BUILD_TYPE "Release" CACHE STRING "CMake build type")
```

Reviewer: IIUC all the currently failing builds are caused by this `CACHE STRING "CMake build type"` option ...

Reviewer 2: Weird.
Let's focus on this github action run: https://github.com/libtom/libtommath/actions/runs/2024483993

All errors are due to missing symbols.

Reviewer 2: Seems like you fixed it somehow?
https://github.com/libtom/libtommath/actions/runs/2024551731

Reviewer: e.g. https://github.com/libtom/libtommath/runs/5650110146?check_suite_focus=true

clang requires at least `-O1` to build library and tests, gcc only requires at least `-O1` for the tests

As soon as I add the `CACHE STRING "CMake build type"` the default `CMAKE_BUILD_TYPE` isn't taken into account, but still stays at `""`. I've just removed it again and that build now succeeded https://github.com/libtom/libtommath/runs/5650310855?check_suite_focus=true

I'll simply leave it out. "It works as it is now"^TM :)

Reviewer: > Seems like you fixed it somehow?

yes, by removing this option

Reviewer 2: You're the libtommath pro here, but are `s_read_wincsp` and `s_read_arc4random` exported symbols by the lbrary?
Isn't the issue here that a consumer is using private symbols which the compiler apparently stripped or does not export from the library?

Reviewer: Those are static functions which should be dead-code-eliminated and the problem is that clang only does that from `-O1` on.

include(GNUInstallDirs)
include(CMakePackageConfigHelpers)
include(sources.cmake)

# The only direct cmake argument for now
option(BUILD_SHARED_LIBS "Build shared library and only the shared library if \"ON\", default is static" OFF)

#-----------------------------------------------------------------------------
# Compose CFLAGS
#-----------------------------------------------------------------------------

# Some information copied from makefile_include.mk

# Basic set
set(LTM_C_FLAGS  -Wall -Wsign-compare -Wextra -Wshadow)
set(LTM_C_FLAGS  ${LTM_C_FLAGS} -Wdeclaration-after-statement -Wbad-function-cast -Wcast-align)
set(LTM_C_FLAGS  ${LTM_C_FLAGS} -Wstrict-prototypes -Wpointer-arith)

# Do we run the sanitizer?
if(DEFINED ENV{SANITIZER})
    set(LTM_C_FLAGS  ${LTM_C_FLAGS} -fsanitize=undefined -fno-sanitize-recover=all -fno-sanitize=float-divide-by-zero)
endif()

if(DEFINED ENV{CONV_WARNINGS})
    set(LTM_C_FLAGS  ${LTM_C_FLAGS} -std=c89 -Wconversion -Wsign-conversion)
    if($ENV{CONV_WARNINGS} EQUAL "strict")
        set(LTM_C_FLAGS  ${LTM_C_FLAGS} -Wc++-compat)
    endif()
else()
    set(LTM_C_FLAGS  ${LTM_C_FLAGS} -Wsystem-headers)

> endif()

Reviewer: Not sure if those are really required here, as this exists only for the unit tests and is already done in the makefiles.

Reviewer 3: Uh, yes, the unit tests use the makefiles, so these lines can go.

        #x86 optimizations [should be valid for any GCC install though]
        set(LTM_C_FLAGS  ${LTM_C_FLAGS} -fomit-frame-pointer)
    endif()
    # TODO:
    # if(DEFINED ENV{COMPILE_LTO})
    #     set(LTM_C_FLAGS  "${LTM_C_FLAGS} -flto")
    #     set(LTM_LD_FLAGS "${LTM_LD_FLAGS} -flto")
    #     #AR = $(subst clang,llvm-ar,$(subst gcc,gcc-ar,$(CC)))
    # endif()
endif()

# TODO
# Are the coming three checks really the best way?

# What compiler do we have and what are their...uhm... peculiarities
# TODO: is the check for a C++ compiler necessary?
if( (CMAKE_C_COMPILER_ID MATCHES "(C|c?)lang") OR (CMAKE_CXX_COMPILER_ID MATCHES "(C|c?)lang"))
    set(LTM_C_FLAGS  ${LTM_C_FLAGS} -Wno-typedef-redefinition -Wno-tautological-compare -Wno-builtin-requires-header)
endif()

if( (CMAKE_C_COMPILER_ID MATCHES "mingw") OR (CMAKE_CXX_COMPILER_ID MATCHES "mingw"))
    set(LTM_C_FLAGS  ${LTM_C_FLAGS}  -Wno-shadow)
endif()

if(DEFINED ENV{PLATFORM})
    if($ENV{PLATFORM} MATCHES "Darwin")
        set(LTM_C_FLAGS  ${LTM_C_FLAGS}  -Wno-nullability-completeness)
    endif()
    if($ENV{PLATFORM} MATCHES "CYGWIN")
        set(LTM_C_FLAGS  ${LTM_C_FLAGS} -no-undefined)
    endif()

> endif()

Reviewer: Is this really the best way to do this?

Reviewer 3: I don't have the faintest idea although my bet is on: probably not. I could only spend about an hour on it before my coffee run out.

RegExp put in compiler checks because it seems as if `CMAKE_C_COMPILER_ID` contains either `Clang` or `clang` under certain circumstances (athough [CMake docs](https://cmake.org/cmake/help/latest/variable/CMAKE_LANG_COMPILER_ID.html) list only `Clang` ) and I found no way to make `MATCH` case insensitive.

There are several CMake build-ins for platform checks, see e.g. [this link](https://gitlab.kitware.com/cmake/community/-/wikis/doc/tutorials/How-To-Write-Platform-Checks) that has a couple of lists at the end of the page with the macros but I found no *exact* definitions for them (but did not dive into CMake's sourcecode).

    $<INSTALL_INTERFACE:${CMAKE_INSTALL_INCLUDEDIR}/${PROJECT_NAME}>
)

target_compile_options(${PROJECT_NAME} BEFORE PRIVATE
    ${LTM_C_FLAGS}
)
target_link_options(${PROJECT_NAME} BEFORE PRIVATE
    ${LTM_LD_FLAGS}
)

set_target_properties(${PROJECT_NAME} PROPERTIES VERSION ${PROJECT_VERSION})
set_target_properties(${PROJECT_NAME} PROPERTIES SOVERSION ${PROJECT_VERSION_MAJOR})

#-----------------------------------------------------------------------------
# demo target
#-----------------------------------------------------------------------------

add_executable(test-target EXCLUDE_FROM_ALL
    ${CMAKE_CURRENT_SOURCE_DIR}/demo/shared.c
    ${CMAKE_CURRENT_SOURCE_DIR}/demo/test.c
)

target_include_directories(test-target PRIVATE
    ${CMAKE_CURRENT_SOURCE_DIR}
)

target_link_libraries(test-target PRIVATE
    ${PROJECT_NAME}
)

target_compile_options(test-target BEFORE PRIVATE
    $<$<STREQUAL:$<TARGET_PROPERTY:${PROJECT_NAME},TYPE>,SHARED_LIBRARY>:-O1 -DLTM_TEST_DYNAMIC>
    ${LTM_C_FLAGS}
)
target_link_options(test-target BEFORE PRIVATE
    $<$<STREQUAL:$<TARGET_PROPERTY:${PROJECT_NAME},TYPE>,SHARED_LIBRARY>:-O1>
    ${LTM_LD_FLAGS}
)

> 

Reviewer: Yeah ... took me only half an afternoon to find out ;)

set(CPACK_PACKAGE_DESCRIPTION_SUMMARY "LibTomMath")
set(CPACK_PACKAGE_VENDOR "LibTomMath")
set(CPACK_PACKAGE_CONTACT "libtom@googlegroups.com")
set(CPACK_RESOURCE_FILE_LICENSE "${PROJECT_SOURCE_DIR}/LICENSE")
set(CPACK_PACKAGE_FILE_NAME ${CPACK_PACKAGE_NAME}-${CPACK_PACKAGE_VERSION}-${PACKAGE_RELEASE_VERSION}_${LINUX_DISTRO}-${LINUX_DISTRO_VERSION}_${MACHINE_ARCH})
set(CPACK_STRIP_FILES ON)

# deb specific CPack config
set(CPACK_DEBIAN_PACKAGE_RELEASE ${PACKAGE_RELEASE_VERSION})
set(CPACK_DEBIAN_FILE_NAME DEB-DEFAULT)
set(CPACK_DEBIAN_PACKAGE_SHLIBDEPS ON)
set(CPACK_DEBIAN_PACKAGE_HOMEPAGE "https://www.libtom.net/LibTomMath")
set(CPACK_DEBIAN_PACKAGE_SECTION "devel")

# rpm specific CPack config
set(CPACK_RPM_PACKAGE_URL "https://www.libtom.net/LibTomMath")
set(CPACK_RPM_PACKAGE_RELEASE ${PACKAGE_RELEASE_VERSION})
set(CPACK_RPM_PACKAGE_ARCHITECTURE ${MACHINE_ARCH})
set(CPACK_RPM_PACKAGE_NAME "lib${PROJECT_NAME}-${PROJECT_VERSION}")
set(CPACK_RPM_FILE_NAME "lib${PROJECT_NAME}_${PROJECT_VERSION}-${CPACK_RPM_PACKAGE_RELEASE}_${LINUX_DISTRO}-${LINUX_DISTRO_VERSION}_${CPACK_RPM_PACKAGE_ARCHITECTURE}.rpm")
set(CPACK_RPM_PACKAGE_LICENSE "WTFPL")

> include(CPack)

Reviewer 3: There is also a [CPack BSD](https://cmake.org/cmake/help/latest/cpack_gen/freebsd.html) for the `pkg` format but it has been added in CMake 3.10

Reviewer: I guess we could up our minimum CMake requirements as 3.7 is pretty old. If we take "everyone's preferred Distro"^TM as reference, Debian Stretch is the only still-maintained version that brings only 3.7, all others bring at least 3.10. Stretch-backports also brings 3.16, so we should be fine with upping to 3.10.

Reviewer: > I guess we could up our minimum CMake requirements as 3.7 is pretty old. If we take "everyone's preferred Distro"^TM as reference, Debian Stretch is the only still-maintained version that brings only 3.7, all others bring at least 3.10. Stretch-backports also brings 3.16, so we should be fine with upping to 3.10.

@czurnieden @mrpelotazo your thoughts on this?

Reviewer 3: Looking at the BSD's: FreeBSD has 3.5 for their oldest  version 11 but only for [selected architectures](https://www.freshports.org/devel/cmake), [OpenBSD has 3.19](https://ftp.openbsd.org/pub/OpenBSD/6.9/packages/) in its oldest version 6.9, NetBSD seems to have the newest with [3.22](http://cvsweb.netbsd.org/bsdweb.cgi/pkgsrc/devel/cmake/). I didn't check the more obscure OSs LTM supports but I think were on the safe side with CMake 3.10

Reviewer 3: Cpack FreeBSD works in FreeBSD 13 (CPack needs to be included manually when building cmake from ports but every FreeBSD developer should know that. I did not ;-) ) although we forgot to add it to the list. See [my fork of this branch](https://github.com/czurnieden/libtommath/tree/add-cmake-support)

Author: @sjaeckel well I don't know how far back you guys want to go. Version 3.10 sounds reasonable to me.

Reviewer: FYI I've added the FreeBSD section to CPack even though it doesn't work... CPack support in FreeBSD's CMake is marked as experimental and after enabling it and re-building cmake, it segfaults if executed ...

```
$ cpack -G "FREEBSD"
CPack: Create package using FREEBSD
CPack: Install projects
CPack: - Run preinstall target for: tommath
CPack: - Install project: tommath []
CPack: Create package
Segmentation fault (core dumped)
```

CC @gahr @lbartoletti

Reviewer 4: Sorry I have no experience with CPath and I haven't followed this discussion. What do we need it for?

Reviewer 4: Oh, I just read it up. Can we turn it off on FreeBSD? We don't use it to generate packages anyway.

    list(APPEND LTM_C_FLAGS -Wno-nullability-completeness)
endif()
if(CMAKE_SYSTEM_NAME MATCHES "CYGWIN")
    list(APPEND LTM_C_FLAGS -no-undefined)
endif()

# TODO: coverage (lgcov)

# If the user set the environment variables at generate-time, append them
# in order to allow overriding our defaults.
# ${LTM_CFLAGS} means the user passed it via sth like:
# $ cmake -DLTM_CFLAGS="foo"
list(APPEND LTM_C_FLAGS ${LTM_CFLAGS})
list(APPEND LTM_LD_FLAGS ${LTM_LDFLAGS})

#-----------------------------------------------------------------------------
# library target
#-----------------------------------------------------------------------------
add_library(${PROJECT_NAME}
    ${SOURCES}
    ${HEADERS}
)

target_include_directories(${PROJECT_NAME} PUBLIC
    $<BUILD_INTERFACE:${CMAKE_CURRENT_SOURCE_DIR}>
    $<INSTALL_INTERFACE:${CMAKE_INSTALL_INCLUDEDIR}/${PROJECT_NAME}>
)

target_compile_options(${PROJECT_NAME} BEFORE PRIVATE
    ${LTM_C_FLAGS}
)
target_link_options(${PROJECT_NAME} BEFORE PRIVATE
    ${LTM_LD_FLAGS}
)

set_target_properties(${PROJECT_NAME} PROPERTIES
    VERSION ${PROJECT_VERSION}
    SOVERSION ${PROJECT_VERSION_MAJOR}
    PUBLIC_HEADER tommath.h
)

> option(COMPILE_LTO "Build with LTO enabled")
> check_ipo_supported(RESULT result)
> if(result AND COMPILE_LTO)
>     set_property(TARGET ${PROJECT_NAME} PROPERTY INTERPROCEDURAL_OPTIMIZATION TRUE)
> elseif(COMPILE_LTO)
>     message(WARNING "LTO is not supported for this compiler.")
> endif()

Reviewer 2: Let's throw an error if a user passes `-DCOMPILE_LTO=ON`, but the compiler does not support LTO.

```suggestion
option(COMPILE_LTO "Build with LTO enabled")
check_ipo_supported(RESULT COMPILER_SUPPORTS_LTO)
if(COMPILER_SUPPORTS_LTO AND COMPILE_LTO)
    set_property(TARGET ${PROJECT_NAME} PROPERTY INTERPROCEDURAL_OPTIMIZATION TRUE)
elseif(COMPILE_LTO AND NOT COMPILER_SUPPORTS_LTO)
    message(SEND_ERROR "This compiler does not support LTO. Reconfigure ${PROJECT_NAME} with -DCOMPILE_LTO=OFF.")
endif()
```

project(tommath
    VERSION 1.2.0
    DESCRIPTION "A free open source portable number theoretic multiple-precision integer (MPI) library written entirely in C."
    HOMEPAGE_URL "https://www.libtom.net/LibTomMath"
    LANGUAGES C)

# package release version
# bump if re-releasing the same VERSION + patches
# set to 1 if releasing a new VERSION
set(PACKAGE_RELEASE_VERSION 1)

#-----------------------------------------------------------------------------
# Include cmake modules
#-----------------------------------------------------------------------------
include(GNUInstallDirs)
include(CMakePackageConfigHelpers)
include(sources.cmake)

# The only direct cmake argument for now
option(BUILD_SHARED_LIBS "Build shared library and only the shared library if \"ON\", default is static" OFF)

#-----------------------------------------------------------------------------
# Compose CFLAGS
#-----------------------------------------------------------------------------

# Some information copied from makefile_include.mk

# Basic set
set(LTM_C_FLAGS  -Wall -Wsign-compare -Wextra -Wshadow)
set(LTM_C_FLAGS  ${LTM_C_FLAGS} -Wdeclaration-after-statement -Wbad-function-cast -Wcast-align)

> set(LTM_C_FLAGS  ${LTM_C_FLAGS} -Wstrict-prototypes -Wpointer-arith -Wsystem-headers)

Reviewer 3: > These are not compatible with Visual Studio.

VS needs a lot of extra massages for LTM, see e.g.: #525 for the details. It is a LUA script and not a very small one. Would it be possible to do it in CMake or would it be too much of a hassle?

> Also, you can use list(APPEND) to append to a list, or do a single set() that spans multiple lines.

Ah, thanks!

if(NOT CMAKE_BUILD_TYPE AND NOT CMAKE_CONFIGURATION_TYPES)
  message(STATUS "Setting build type to 'Release' as none was specified.")

>   set(CMAKE_BUILD_TYPE "Release")

Reviewer: What should we do here? I didn't add the `CACHE...` option as it doesn't seem to work ... maybe you wanna have another look?

>   set(CMAKE_BUILD_TYPE "Release")

Reviewer: NVM, I simply added it ...

endif()

set(CMAKE_C_FLAGS_DEBUG "-g3")
set(CMAKE_C_FLAGS_RELEASE "-O3 -funroll-loops -fomit-frame-pointer")
set(CMAKE_C_FLAGS_RELWITHDEBINFO "-g3 -O2")
set(CMAKE_C_FLAGS_MINSIZEREL "-Os")

> if(COMPILE_LTO)

Reviewer 3: Ah, it is fully supported now (since CMake 3.9 if I understood it correctly, we had an older version at the beginning of this branch which only supported Intel's compiler, or so it seemed). Good to know and easy to implement.

    set(LTM_C_FLAGS  ${LTM_C_FLAGS} -flto)
    set(LTM_LD_FLAGS ${LTM_LD_FLAGS} -flto)
endif()

# What compiler do we have and what are their...uhm... peculiarities
if(CMAKE_C_COMPILER_ID MATCHES "(C|c?)lang")

>     set(LTM_C_FLAGS  ${LTM_C_FLAGS} -Wno-typedef-redefinition -Wno-tautological-compare -Wno-builtin-requires-header)

Author: Indeed

    # Clang requires at least '-O1' for dead code eliminiation
    set(CMAKE_C_FLAGS_DEBUG  "${CMAKE_C_FLAGS_DEBUG} -O1")
endif()
if(CMAKE_C_COMPILER MATCHES "mingw")

>     set(LTM_C_FLAGS  ${LTM_C_FLAGS}  -Wno-shadow)

Author: ```suggestion
    list(APPEND LTM_C_FLAGS  -Wno-shadow)
```

endif()
if(CMAKE_SYSTEM_NAME MATCHES "Darwin")

>     set(LTM_C_FLAGS  ${LTM_C_FLAGS}  -Wno-nullability-completeness)

Author: ```suggestion
    list(APPEND LTM_C_FLAGS  -Wno-nullability-completeness)
```

endif()
if(CMAKE_SYSTEM_NAME MATCHES "CYGWIN")

>     set(LTM_C_FLAGS  ${LTM_C_FLAGS} -no-undefined)

Author: ```suggestion
    list(APPEND LTM_C_FLAGS -no-undefined)
```

endif()

> # TODO: coverage (lgcov)

Reviewer 3: Probable license problems (They use BSD 3-clause which is not compatible with ours).

# If the user set the environment variables at generate-time, append them
# in order to allow overriding our defaults.
# ${LTM_CFLAGS} means the user passed it via sth like:
# $ cmake -DLTM_CFLAGS="foo"

> set(LTM_C_FLAGS  ${LTM_C_FLAGS} ${LTM_CFLAGS})

Author: ```suggestion
list(APPEND LTM_C_FLAGS ${LTM_CFLAGS})
```

> set(LTM_LD_FLAGS  ${LTM_LD_FLAGS} ${LTM_LDFLAGS})

Author: ```suggestion
list(APPEND LTM_LD_FLAGS ${LTM_LDFLAGS})
```

#-----------------------------------------------------------------------------
# library target
#-----------------------------------------------------------------------------
add_library(${PROJECT_NAME}
    ${SOURCES}
)

target_include_directories(${PROJECT_NAME} PUBLIC
    $<BUILD_INTERFACE:${CMAKE_CURRENT_SOURCE_DIR}>
    $<INSTALL_INTERFACE:${CMAKE_INSTALL_INCLUDEDIR}/${PROJECT_NAME}>
)

target_compile_options(${PROJECT_NAME} BEFORE PRIVATE
    ${LTM_C_FLAGS}
)
target_link_options(${PROJECT_NAME} BEFORE PRIVATE
    ${LTM_LD_FLAGS}
)

set_target_properties(${PROJECT_NAME} PROPERTIES
    VERSION ${PROJECT_VERSION}
    SOVERSION ${PROJECT_VERSION_MAJOR}
)

> 

Author: Set PUBLIC_HEADER property so the install with `PUBLIC_HEADER DESTINATION` picks it.
```suggestion
set_target_properties(${PROJECT_NAME} PROPERTIES PUBLIC_HEADER
    ${PUBLIC_HEADERS}
)
```

#-----------------------------------------------------------------------------
# demo target
#-----------------------------------------------------------------------------

add_executable(test-target EXCLUDE_FROM_ALL
    ${CMAKE_CURRENT_SOURCE_DIR}/demo/shared.c
    ${CMAKE_CURRENT_SOURCE_DIR}/demo/test.c
)

target_include_directories(test-target PRIVATE
    ${CMAKE_CURRENT_SOURCE_DIR}
)

target_link_libraries(test-target PRIVATE
    ${PROJECT_NAME}
)

target_compile_options(test-target BEFORE PRIVATE

>     $<$<STREQUAL:$<TARGET_PROPERTY:${PROJECT_NAME},TYPE>,SHARED_LIBRARY>:-O1 -DLTM_TEST_DYNAMIC>

Reviewer 3: > Lowering the optimization level to -O1 might hide bugs caused by additional optimizations.

Anybody remember what the exact reasons were for this restriction?

>     $<$<STREQUAL:$<TARGET_PROPERTY:${PROJECT_NAME},TYPE>,SHARED_LIBRARY>:-O1 -DLTM_TEST_DYNAMIC>

Reviewer: > Why do you override the optimization level here and below?
People should be able build/run the tests in debug/relase/... mode.

In the versions before we didn't use the CMake specific modes, but only `LTM_C_FLAGS` and so this `-O1` was overridden by the user when `LTM_C_FLAGS` was appended.

Now we use those CMake specific variables and there's no way to prepend something ...

> There is target_compile_definitions to add compile definitions.

this doesn't help, since we still have to make sure to have at least `-O1`

    ${LTM_C_FLAGS}
)
target_link_options(test-target BEFORE PRIVATE
    $<$<STREQUAL:$<TARGET_PROPERTY:${PROJECT_NAME},TYPE>,SHARED_LIBRARY>:-O1>
    ${LTM_LD_FLAGS}
)

#-----------------------------------------------------------------------------
# demo target
#-----------------------------------------------------------------------------
add_custom_target(check
    COMMAND ${CMAKE_CURRENT_BINARY_DIR}/test-target
)

> add_dependencies(check test-target)

Author: I'd enable CTest as documented [here](https://cmake.org/cmake/help/v3.10/module/CTest.html?highlight=build_testing#ctest).

#-----------------------------------------------------------------------------
# Install/export targets and files
#-----------------------------------------------------------------------------
set(CONFIG_INSTALL_DIR "${CMAKE_INSTALL_LIBDIR}/cmake/${PROJECT_NAME}")
set(PROJECT_VERSION_FILE "${CMAKE_CURRENT_BINARY_DIR}/${PROJECT_NAME}-config-version.cmake")
set(PROJECT_CONFIG_FILE "${PROJECT_NAME}-config.cmake")
set(TARGETS_EXPORT_NAME "${PROJECT_NAME}Targets")

install(TARGETS ${PROJECT_NAME}
    EXPORT ${TARGETS_EXPORT_NAME}
    LIBRARY DESTINATION ${CMAKE_INSTALL_LIBDIR}

>     ARCHIVE DESTINATION ${CMAKE_INSTALL_LIBDIR}

Author: Use install PUBLIC_HEADER instead of install(FILES ${HEADERS})
```suggestion
    ARCHIVE DESTINATION ${CMAKE_INSTALL_LIBDIR}
    PUBLIC_HEADER DESTINATION ${CMAKE_INSTALL_INCLUDEDIR}/${PROJECT_NAME}
```

)

> install(FILES ${HEADERS}
>     DESTINATION ${CMAKE_INSTALL_INCLUDEDIR}/${PROJECT_NAME}
> )

Author: Use install PUBLIC_HEADER instead of install(FILES ${HEADERS})
```suggestion
```

# Install libtommath.pc for pkg-config if we build a shared library
if(BUILD_SHARED_LIBS)
    configure_file(
        ${CMAKE_CURRENT_SOURCE_DIR}/lib${PROJECT_NAME}.pc.in
        ${CMAKE_CURRENT_BINARY_DIR}/lib${PROJECT_NAME}.pc
        @ONLY
    )

    install(FILES ${CMAKE_CURRENT_BINARY_DIR}/lib${PROJECT_NAME}.pc

>         DESTINATION ${CMAKE_INSTALL_DATAROOTDIR}/pkgconfig

Author: For what I see in my system you're right. Most of the .pc files are located at `${CMAKE_INSTALL_LIBDIR}/pkgconfig`

>         DESTINATION ${CMAKE_INSTALL_DATAROOTDIR}/pkgconfig

Reviewer 3: Since version 0.24 `pkgconfig` returns the paths with `pkg-config --variable pc_path pkg-config`

```
$ lsb_release -a
LSB Version:	core-11.1.0ubuntu2-noarch:security-11.1.0ubuntu2-noarch
Distributor ID:	Linuxmint
Description:	Linux Mint 20.3
Release:	20.3
Codename:	una`
$ pkg-config --version
0.29.1
$ pkg-config --variable pc_path pkg-config
/usr/local/lib/x86_64-linux-gnu/pkgconfig:/usr/local/lib/pkgconfig:/usr/local/share/pkgconfig:/usr/lib/x86_64-linux-gnu/pkgconfig:/usr/lib/pkgconfig:/usr/share/pkgconfig
```

Can we make use of it? Does CMake make use of it automatically? Or just use` ${CMAKE_INSTALL_LIBDIR}` as proposed?

>         DESTINATION ${CMAKE_INSTALL_DATAROOTDIR}/pkgconfig

Reviewer 2: Applying your suggestion would require:
1. look for the presenceo of `pkg-config` ([`FIndPkgConfig`](https://cmake.org/cmake/help/latest/module/FindPkgConfig.html))
2. Run `PKG_CONFIG_EXECUTABLE --variable pc_path pkg-config`
3. Parse the output

In any cases, we would need to supply a default in case pkg-config is not available.
This also introduces some level of uncertainty and dependency on pkg-config.

How about the following?
```cmake
include(GNUInstallDirs)
set(CMAKE_INSTALL_PKGCONFIGDIR "${CMAKE_INSTALL_LIBDIR}/pkgconfig" CACHE PATH "Folder where to install .pc files")
```
This still allows people to override it.
This is simple and 100% reproducible.

>         DESTINATION ${CMAKE_INSTALL_DATAROOTDIR}/pkgconfig

Reviewer 3: > This is not very hard, but also not trivial in cmake.

Yepp, feared that. But you're right: let's keep it simple.

    )
endif()

# generate package version file
write_basic_package_version_file(
    ${PROJECT_VERSION_FILE}
    VERSION ${PROJECT_VERSION}
    COMPATIBILITY SameMajorVersion
)

# install version file
install(FILES ${PROJECT_VERSION_FILE}
        DESTINATION ${CONFIG_INSTALL_DIR}
)

# build directory package config
export(EXPORT ${TARGETS_EXPORT_NAME}
       FILE ${PROJECT_CONFIG_FILE}
)

# installed package config
install(EXPORT ${TARGETS_EXPORT_NAME}
        DESTINATION ${CONFIG_INSTALL_DIR}
        FILE ${PROJECT_CONFIG_FILE}
)

# add to CMake registry
export(PACKAGE ${PROJECT_NAME})

#---------------------------------------------------------------------------------------

> # Create release packages

Reviewer: thx but meh ... we have `make zipup` :)

#---------------------------------------------------------------------------------------

# determine distribution and architecture
find_program(LSB_RELEASE lsb_release)

> execute_process(COMMAND uname -m OUTPUT_VARIABLE MACHINE_ARCH OUTPUT_STRIP_TRAILING_WHITESPACE)

Author: You're probably right. I'd rather use [CMAKE_HOST_SYSTEM_NAME](https://cmake.org/cmake/help/v3.10/variable/CMAKE_HOST_SYSTEM_NAME.html#cmake-host-system-name) though.

> execute_process(COMMAND uname -m OUTPUT_VARIABLE MACHINE_ARCH OUTPUT_STRIP_TRAILING_WHITESPACE)

Reviewer 2: `CMAKE_HOST_SYSTEM_NAME` is the name of the system running cmake (the host system for cmake).
`CMAKE_SYSTEM_NAME` is the name of the system we are building for.

So `CMAKE_HOST_SYSTEM_NAME` is the build system name,
and `CMAKE_SYSTEM_NAME` is the host system name.

(The naming is counter-intuitive)

> execute_process(COMMAND uname -m OUTPUT_VARIABLE MACHINE_ARCH OUTPUT_STRIP_TRAILING_WHITESPACE)

Author: They hold host and target. A simple CMakeLists.txt just printing these variables:

Building on linux for linux:
```
cmake ..
...
CMAKE_HOST_SYSTEM_NAME: Linux
CMAKE_SYSTEM_NAME     : Linux
...
```

Building on linux for windows:
```
x86_64-w64-mingw32-cmake ..
...
CMAKE_HOST_SYSTEM_NAME: Linux
CMAKE_SYSTEM_NAME     : Windows
...
```

> execute_process(COMMAND uname -m OUTPUT_VARIABLE MACHINE_ARCH OUTPUT_STRIP_TRAILING_WHITESPACE)

Reviewer 2: See https://crosstool-ng.github.io/docs/toolchain-types/
There exists build, host and target.
Since libtommath is not a compiler, host == target.

> execute_process(COMMAND uname -m OUTPUT_VARIABLE MACHINE_ARCH OUTPUT_STRIP_TRAILING_WHITESPACE)

Author: The crosstool link may be a little confusing since it throws toolchain building in the mix.

We are just running a toolchain to build the lib.

So we are left with two questions:

1. In what system is the toolchain running i.e. the system where the lib is built (host)
2. In what system is the built lib going to run (target)

Bullet point 1. from the link matches the first example above:
    Build (toolchain) runs under linux and produces a lib to be run on linux.
    Thus host=linux target=linux

Bullet point 2.  from the link matches the second one:
    Build (toolchain) runs under linux and produces a lib to be run on windows.
    Thus host=linux target=windows

So cmake variable CMAKE_HOST_SYSTEM_NAME is the host and CMAKE_SYSTEM_NAME is the target.

Hope this clarifies it.

> execute_process(COMMAND uname -m OUTPUT_VARIABLE MACHINE_ARCH OUTPUT_STRIP_TRAILING_WHITESPACE)

Reviewer: > So cmake variable CMAKE_HOST_SYSTEM_NAME is the host and CMAKE_SYSTEM_NAME is the target and they are not necessarily the same.

:+1: 

In those parts of the file we're defining how the cpack file is called, so the name shall be derived from `CMAKE_SYSTEM_NAME`.

btw. using regular CMake with the toolchain file provided by [`mingw-w64-cmake`](https://aur.archlinux.org/packages/mingw-w64-cmake) produces a functional binary.

```
$ cmake --toolchain=/usr/share/mingw/toolchain-x86_64-w64-mingw32.cmake ..
...
-- CMAKE_HOST_SYSTEM_NAME="Linux"
-- CMAKE_SYSTEM_NAME="Windows"
-- CMAKE_EXECUTABLE_SUFFIX=".exe"
-- MINGW="1"
...
$ make -j65 check
...
Tests OK/NOP/FAIL: 47/0/0
[100%] Built target check
$ file test-target*
test-target.exe: PE32+ executable (console) x86-64, for MS Windows
```

Using [`mingw-w64-cmake`](https://aur.archlinux.org/packages/mingw-w64-cmake) itself produces a broken binary ... (at least for Wine, I don't have Windows to test) ... maybe it's also the AUR build that is broken ...

```
$ x86_64-w64-mingw32-cmake ..
...
-- CMAKE_HOST_SYSTEM_NAME="Linux"
-- CMAKE_SYSTEM_NAME="Windows"
-- CMAKE_EXECUTABLE_SUFFIX=".exe"
$ make check
...
[100%] Linking C executable test-target.exe
[100%] Built target test-target
0100:err:module:import_dll Library libgcc_s_seh-1.dll (which is needed by L"Z:\\path\\to\\libtommath\\build\\libtommath.dll") not found
0100:err:module:import_dll Library libtommath.dll (which is needed by L"Z:\\path\\to\\libtommath\\build\\test-target.exe") not found
0100:err:module:LdrInitializeThunk Importing dlls for L"Z:\\path\\to\\libtommath\\build\\test-target.exe" failed, status c0000135
make[3]: *** [CMakeFiles/check.dir/build.make:70: CMakeFiles/check] Error 53
make[2]: *** [CMakeFiles/Makefile2:137: CMakeFiles/check.dir/all] Error 2
make[1]: *** [CMakeFiles/Makefile2:144: CMakeFiles/check.dir/rule] Error 2
make: *** [Makefile:215: check] Error 2
$ file test-target*
test-target.exe: PE32+ executable (console) x86-64, for MS Windows
```

> execute_process(COMMAND uname -m OUTPUT_VARIABLE MACHINE_ARCH OUTPUT_STRIP_TRAILING_WHITESPACE)

Author: I could build and run `test-target.exe` within a mingw docker container based on [this one](https://github.com/General-Beck/docker-mingw-qt5) I use for cross building Qt stuff for windows.

I still get an error though due to a path issue:
`make[3]: ./test-target: No such file or directory`

But manually running `test-target.exe` works [test-target.log](https://github.com/libtom/libtommath/files/8314137/test-target.log).

@sjaeckel in theory invoking `x86_64-w64-mingw32-cmake` directly should pick the right toolchain file. This is e.g. the `ccmake` output from within the build directory (NOTE the CMAKE_TOOLCHAIN_FILE variable):

```
 BUILD_SHARED_LIBS                ON
 CMAKE_BUILD_TYPE                 
 CMAKE_CXX_IMPLICIT_INCLUDE_DIR   /usr/x86_64-w64-mingw32/include
 CMAKE_C_IMPLICIT_INCLUDE_DIREC   /usr/x86_64-w64-mingw32/include
 CMAKE_GNUtoMS                    OFF
 CMAKE_INSTALL_PREFIX             /usr/x86_64-w64-mingw32
 CMAKE_TOOLCHAIN_FILE             /usr/share/mingw/toolchain-x86_64-w64-mingw32.cmake
 LSB_RELEASE                      /usr/sbin/lsb_release
```

> execute_process(COMMAND uname -m OUTPUT_VARIABLE MACHINE_ARCH OUTPUT_STRIP_TRAILING_WHITESPACE)

Reviewer: > I still get an error though due to a path issue:

Yeah, I've still to commit&push all my local changes ...

```
$ git diff | grep -B1 EXE
-    COMMAND ${CMAKE_CURRENT_BINARY_DIR}/test-target
+    COMMAND ${CMAKE_CURRENT_BINARY_DIR}/${LTM_TEST}${CMAKE_EXECUTABLE_SUFFIX}
```
